src/junit.rs

```diff
@@ -10,36 +10,42 @@ use quick_xml::reader::Reader;
 use crate::testrun::{Outcome, Testrun};
 
 // from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
-fn attributes_map(attributes: Attributes) -> HashMap<String, String> {
-    return attributes
-        .filter_map(|a| {
-            if let Ok(attr) = a {
-                let bytes = attr.value.into_owned();
-                let value = String::from_utf8(bytes).unwrap();
-                let key = String::from_utf8(attr.key.local_name().as_ref().to_vec()).unwrap();
-                Some((key, value))
-            } else {
-                None
-            }
-        })
-        .collect::<HashMap<_, _>>();
+fn attributes_map(attributes: Attributes) -> Result<HashMap<String, String>, pyo3::PyErr> {
```

Comment:
Wondering what's your thought process in a) moving from closures to a less functional programming approach and b) explicitly returning Result - seems to be proper error handling

I'm getting a sense Result, returning Ok or Err, is analogous to Try/Catch ya?

Reply:
a) I just couldn't figure out how the `?` operator worked within closures, it was complicating thinking about that so I just moved everything into high level for loops to simplify it, I would've liked to keep the closures but I need to figure out how Result interacts with closures and iterators before that

b) Yeah, I want to be able to propagate errors back to the high level library function we're exposing

I guess it kind of is like try/catch, but instead of having statements you have an actual object, I actually find this nicer than try/catch specifically because of how enums can contain values in Rust

Reply:
Yup, enums are different in Rust, cool ty!